nntrainer/models/neuralnet.cpp

```diff
@@ -478,8 +482,8 @@ bool NeuralNetwork::validateInput(sharedConstTensors X) {
   return true;
 }
 
-sharedConstTensors NeuralNetwork::inference(sharedConstTensors X,
-                                            bool free_mem) {
+std::vector<sharedConstTensors> NeuralNetwork::inference(sharedConstTensors X,
```

Comment:
Currently, this is returning vector of vector.
The reasoning behind this is that each node can have multiple outputs.
eg)
out_layer1 -> out_tensor_a, out_tensor_b
out_layer2 -> out_tensor_c, out_tensor_d
out_layer3 -> out_tensor_e,

so returning [[out_tensor_a, out_tensor_b], [out_tensor_c, out_tensor_d], [out_tensor_e]]

We made a bit of offline discussion about this signature and rollback.

Reasoning,
1. NNStreamer or other frameworks expects/makes output of 1d list. (correct me if I am wrong)
2. If we start to allow nested output, not only 2d but nd list is possible. eg) multiple subgraph of above example
3. When receiving inputs, we are having 1D list (float **) not (float ***) so it is also predictable that output is also a type of 1D list.
3. Even though returning 1D result, we are not loosing much representational power because user will know about the output layout.

-------------- Off-the-topic issue
As our sort order is not determined, it will be requested to user to write the out layer names in order, to determine the output layout.

Reply:
> As our sort order is not determined, it will be requested to user to write the out layer names in order, to determine the output layout.

If these names are to be taken, it better to take them during the model compile. In this way, the shape and size of every inference remains fixed.

Reply:
> We made a bit of offline discussion about this signature and rollback.

Keeping the current signature must be clarified - does it mean multiple output nodes are allowed when the output of each node has to be just 1 tensor or a single node with multiple outputs? IMO, the former case is better, as this can be expanded by making the tensor class more flexible (supporting more than 4 dimensions in the future).

> NNStreamer or other frameworks expects/makes output of 1d list. (correct me if I am wrong)

You are right. However IMO nntrainer's tensor filter element can define semantics for the conversion from nntrainer's inference to be returned to nnstreamer.

> If we start to allow nested output, not only 2d but nd list is possible. eg) multiple subgraph of above example

Can this be nd list (as long as a node can only handle list of tensors)?

Reply:
>  IMO, the former case is better 

I agree :)

> Can this be nd list (as long as a node can only handle list of tensors)?

Yes,

for example.

subgraph 1. [[a], [b,c], [d,e]]
subgarph 2. [[f], [g, h]

output of sugraph 1 + subgraph 2 is 3D list.

This implies that we can't fully represent the list of list with nd graph where n is a fixed number. So the best we can do would be have semantics that is is always flattened for the model output.

Reply:
reverted to return sharedtensors

Reply:
great.